group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java

```diff
@@ -309,53 +326,28 @@ public void commitUpTo(long offset) {
     }
 
     /**
-     * Called when a classic group's state has changed. Increment/decrement
-     * the counter accordingly.
+     * The method is called every {@code classicGroupGaugesUpdateIntervalMs}.
+     * It updates the classic group gauges based on the soft state of the groups.
      *
-     * @param oldState The previous state. null value means that it's a new group.
-     * @param newState The next state. null value means that the group has been removed.
+     * @param groups The classic and consumer groups keyed by their name.
      */
-    public void onClassicGroupStateTransition(
-        ClassicGroupState oldState,
-        ClassicGroupState newState
+    public void updateClassicGroupGauges(
+        TimelineHashMap<String, Group> groups
```

Comment:
I am not really comfortable with passing all the groups to the metrics shard because it would be easy to misused it. Would it be possible to pass `Map<ClassicGroupState, Long>` and to change `classicGroupGauges` to be `volatile`? This way, we could compute the metrics before and just set them here.